crates/adapters/src/test/websocket.rs

```diff
@@ -0,0 +1,152 @@
+//! Helper functions for testing websocket-based communication.
+
+use crate::test::TestStruct;
+use actix_codec::Framed;
+use actix_http::ws::{Codec as WsCodec, Frame as WsFrame, Item as WsItem, Message as WsMessage};
+use byteorder::{BigEndian, ByteOrder};
+use bytes::{BufMut, BytesMut};
+use csv::ReaderBuilder as CsvReaderBuilder;
+use csv::WriterBuilder as CsvWriterBuilder;
+use futures::{SinkExt, StreamExt};
+use std::pin::Pin;
+use tokio::io::{AsyncRead, AsyncWrite};
+
+pub struct TestWsSender;
+
+impl TestWsSender {
+    /// Serialize `data` as `csv` and send it through web socket using one
+    /// single-frame message per buffer.
+    pub async fn send_to_websocket(
+        mut ws: Pin<&mut Framed<impl AsyncRead + AsyncWrite, WsCodec>>,
+        data: &[Vec<TestStruct>],
+    ) {
+        for batch in data.iter() {
+            let mut writer = CsvWriterBuilder::new()
+                .has_headers(false)
+                .from_writer(Vec::with_capacity(batch.len() * 32));
+
+            for val in batch.iter().cloned() {
+                writer.serialize(val).unwrap();
+            }
+            writer.flush().unwrap();
+            let bytes = writer.into_inner().unwrap();
+
+            let mut buf = BytesMut::with_capacity(8 + bytes.len());
+            buf.put(&b"12345678"[..]);
+            buf.put(bytes.as_slice());
+
+            ws.send(WsMessage::Binary(buf.freeze())).await.unwrap();
+            let ack = ws.next().await.unwrap().unwrap();
+            if let WsFrame::Binary(ack) = ack {
+                assert_eq!(ack.len(), 16);
+                assert_eq!(ack[0..8], b"12345678"[..]);
+                assert_eq!(BigEndian::read_u64(&ack[8..16]), (bytes.len() + 8) as u64);
+            } else {
+                panic!("Unexpected websocket response: {ack:?}");
+            }
+        }
+    }
+
+    /// Serialize `data` as `csv` and send it through web socket as one big
+    /// composite message with one frame per buffer.
+    pub async fn send_to_websocket_continuations(
+        mut ws: Pin<&mut Framed<impl AsyncRead + AsyncWrite, WsCodec>>,
+        data: &[Vec<TestStruct>],
+    ) {
+        if data.len() <= 1 {
+            return Self::send_to_websocket(ws, data).await;
```

Comment:
why does this need to be a special case?

Reply:
To break message up into frames in the websocket protocol, one needs at least two frames (first and last).